src/libostree/ostree-repo.c

```diff
@@ -156,6 +156,531 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
 
 #define SYSCONF_REMOTES SHORTENED_SYSCONFDIR "/ostree/remotes.d"
 
+/* Repository locking
+ *
+ * To guard against objects being deleted (e.g., prune) while they're in
+ * use by another operation is accessing them (e.g., commit), the
+ * repository must be locked by concurrent writers.
+ *
+ * The locking is implemented by maintaining a thread local table of
+ * lock stacks per repository. This allows thread safe locking since
+ * each thread maintains its own lock stack. See the OstreeRepoLock type
+ * below.
+ *
+ * The actual locking is done using either open file descriptor locks or
+ * flock locks. This allows the locking to work with concurrent
+ * processes. The lock file is held on the ".lock" file within the
+ * repository.
+ *
+ * The intended usage is to take a shared lock when writing objects or
+ * reading objects in critical sections. Exclusive locks are taken when
+ * deleting objects.
+ *
+ * To allow fine grained locking within libostree, the lock is
+ * maintained as a stack. The core APIs then push or pop from the stack.
+ * When pushing or popping a lock state identical to the existing or
+ * next state, the stack is simply updated. Only when upgrading or
+ * downgrading the lock (changing to/from unlocked, pushing exclusive on
+ * shared or popping exclusive to shared) are actual locking operations
+ * performed.
+ */
+
+static void
+free_repo_lock_table (gpointer data)
+{
+  GHashTable *lock_table = data;
+
+  if (lock_table != NULL)
+    {
+      g_debug ("Free lock table");
+      g_hash_table_destroy (lock_table);
+    }
+}
+
+static GPrivate repo_lock_table = G_PRIVATE_INIT (free_repo_lock_table);
+
+typedef struct {
+  int fd;
+  GQueue *stack;
+} OstreeRepoLock;
+
+typedef struct {
+  guint len;
+  int state;
+  const char *name;
+} OstreeRepoLockInfo;
+
+static void
+repo_lock_info (OstreeRepoLock *lock, OstreeRepoLockInfo *out_info)
+{
+  g_assert (lock != NULL);
+  g_assert (out_info != NULL);
+
+  OstreeRepoLockInfo info;
+  if (lock->stack != NULL)
+    {
+      info.len = g_queue_get_length (lock->stack);
+      if (info.len == 0)
+        {
+          info.state = LOCK_UN;
+          info.name = "unlocked";
+        }
+      else
+        {
+          info.state = GPOINTER_TO_INT (g_queue_peek_head (lock->stack));
+          info.name = (info.state == LOCK_EX) ? "exclusive" : "shared";
+        }
+    }
+  else
+    {
+      info.len = 0;
+      info.state = 0;
+      info.name = "null";
+    }
+
+  *out_info = info;
+}
+
+static void
+free_repo_lock (gpointer data)
+{
+  OstreeRepoLock *lock = data;
+
+  if (lock != NULL)
+    {
+      OstreeRepoLockInfo info;
+      repo_lock_info (lock, &info);
+
+      g_debug ("Free lock: state=%s, depth=%u", info.name, info.len);
+      g_queue_free (lock->stack);
+      if (lock->fd >= 0)
+        {
+          g_debug ("Closing repo lock file");
+          (void) close (lock->fd);
```

Comment:
Do we want to unlink the lock file at any point, for cleanliness?

Although I can’t think of any way that can be race-free, so probably best avoided.

Reply:
IMO, you never want to unlink the lock file since it breaks the semantics of file descriptor locking. At the kernel level, the lock can only be managed if all the callers are operating on the same inode. If you unlink it, then another process trying to acquire the lock would then lock a different inode and the lock is useless. I can show you a concrete example we hit in the Endless jenkins scripts with flock if you'd like.

GLnxLockFile does unlink, but it then goes to great lengths when acquiring the lock to make sure that the file hasn't been unlinked and loops until that's not the case. I don't think that complexity outweighs the simple "don't delete the lock file" advice.

Reply:
> IMO, you never want to unlink the lock file since it breaks the semantics of file descriptor locking. At the kernel level, the lock can only be managed if all the callers are operating on the same inode. If you unlink it, then another process trying to acquire the lock would then lock a different inode and the lock is useless. I can show you a concrete example we hit in the Endless jenkins scripts with flock if you'd like.

Precisely.

> GLnxLockFile does unlink, but it then goes to great lengths when acquiring the lock to make sure that the file hasn't been unlinked and loops until that's not the case. I don't think that complexity outweighs the simple "don't delete the lock file" advice.

I agree with you.